### ISSUE TITLE
Inconsistent Behavior of Data Preparation Pipeline

### ISSUE THREAD

**Issue by akjindal53244** [axolotl-ai-cloud/axolotl#659]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Each record in training dataset should be exactly same upon running the pipeline multiple times.

### Current behaviour

Running the pipeline multiple times gives different input at times.

**Run-1**
<img width="1644" alt="Screenshot 2023-09-30 at 9 27 26 PM" src="https://github.com/OpenAccess-AI-Collective/axolotl/assets/5215386/77b783dc-c9e8-4e9b-9936-13287ba09410">

**Run-2**
<img width="1644" alt="Screenshot 2023-09-30 at 9 30 30 PM" src="https://github.com/OpenAccess-AI-Collective/axolotl/assets/5215386/e3454f48-7461-4e4b-9397-3bc927bb4776">

Notice that tokens for `[INST]` (first word from input) got removed in second run.

### Steps to reproduce

1. Changes to load [dolly-15k dataset](https://huggingface.co/datasets/databricks/databricks-dolly-15k) in ` src/axolotl/prompt_strategies/context_qa.py`:
```
def load_dolly(tokenizer, cfg):
    return ContextQaDollyPromptTokenizingStrategy(
        ContextDollyPrompter(),
        tokenizer,
        cfg.train_on_inputs,
        cfg.sequence_len,
    )


class ContextQaDollyPromptTokenizingStrategy(InstructionPromptTokenizingStrategy):
    """
    Tokenization Strategy to combine in-context article with a question and answer
    """

    def parse_instruction_fields(self, prompt) -> Tuple[str, str, str]:
        instruction = f"[INST] {prompt['instruction']}"
        context = f"Here's some context: {prompt['context']}" if len(prompt["context"]) > 0 else None
        response = f"[/INST] {prompt['response']}"

        if context is not None:
            instruction += f"\n\n{context}"

        return (
            f"{instruction}\n\n",
            "",
            response,
        )


class ContextV2Prompter(AlpacaPrompter):
    """
    Customized system prompted for concise QA
    """

    system_prompt = ""
    system_no_input_prompt = ""

    def match_prompt_style(self):
        # pylint: disable=duplicate-code
        self.turn_format = "{instruction}\n{input}"
        self.turn_no_input_format = "{instruction}"
        self.system_format = "{system}"

```
2. ` examples/llama-2/qlora.yml` file content:
```
base_model: meta-llama/Llama-2-13b-hf
base_model_config: meta-llama/Llama-2-13b-hf
model_type: LlamaForCausalLM
tokenizer_type: LlamaTokenizer
is_llama_derived_model: true

torch_compile:


load_in_8bit: false
load_in_4bit: true
strict: false

datasets:
  - path: databricks/databricks-dolly-15k
    type: context_qa.load_dolly
dataset_prepared_path: ./my_data
val_set_size: 0.01
output_dir: ./qlora-out

adapter: qlora
lora_model_dir:

sequence_len: 2048
sample_packing: false
pad_to_sequence_len: false

lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
lora_target_modules:
    - q_proj
    - v_proj
lora_target_linear:
lora_fan_in_fan_out:

wandb_project:
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:

gradient_accumulation_steps: 32
micro_batch_size: 4
num_epochs: 3
optimizer: paged_adamw_32bit
lr_scheduler: cosine
lr_scheduler_kwargs:
learning_rate: 0.0002

# adamw hyperparams
adam_beta1:
adam_beta2:
adam_epsilon:

# Gradient clipping max norm
max_grad_norm: 0.3

train_on_inputs: false
group_by_length: true
bf16: true
fp16: false
tf32: false

# OOM if disabled
gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 0.03
eval_steps: 20
eval_table_size:
save_steps:
debug: true

weight_decay: 0.001

fsdp:
fsdp_config:

deepspeed: deepspeed/zero2.json

special_tokens:

# Seed
seed: 42
```

3. set `debug_num_examples=1` in `src/axolotl/common/cli.py`
4. Replace `random.randrange(0, len(train_dataset) - 1)` with `0` [here](https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/src/axolotl/cli/__init__.py#L231) so we always print same record in dataset when running pipeline.
5. Disable training by temporarily adding `exit(0)` [here](https://github.com/OpenAccess-AI-Collective/axolotl/blob/main/src/axolotl/train.py#L53).
6. Run it few times to see the difference in output (dataset will be prepared every time from scratch): `rm -rf my_data && accelerate launch -m axolotl.cli.train examples/llama-2/qlora.yml`. I had to run 5 times to see the inconsistent behavior.

### Possible solution

I am new to the repo so need to dig deeper for potential fix.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.9

### axolotl branch-commit

master branch. Commit: 5b0bc48

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Since dolly is alpaca format, could you try this with `alpaca` format and see if same behavior is exhibited?

**Comment by NanoCode012** [comment]
We have added tests for tokenization process a while back. This should prevent any regression here.
